### DISCUSSION TITLE
Automating macOS application icon creation

### DISCUSSION THREAD

**Discussion by aelij** [AvaloniaUI/Avalonia#13756] [Ideas]
Using the `net8.0-macos` TFM creates a macOS app package for Avalonia. However in order to create an icon one has to manually create the images in multiple sizes and run a tool to create a `.icns` file. MAUI has a way to create an app icon from SVG, including all the necessary PNGs and `.plist` updates. I've managed to make both of them work together, so I thought you may want to publish this as a NuGet to assist others.

Caveat - is is a bit hacky.

1. Reference the `Microsoft.Maui.Resizetizer` package.
2. Set the package icon to an SVG file:
    ```xml
    <ItemGroup>
       <MauiIcon Include="../../docs/roslynpad.svg" />
    </ItemGroup>
    ```
3. Set a few properties - note the `.appiconset` path should match the SVG filename:
    ```xml
      <PropertyGroup>
        <_ResizetizerIsiOSApp>true</_ResizetizerIsiOSApp>
        <ResizetizerIncludeSelfProject>true</ResizetizerIncludeSelfProject>
        <OptimizePNGs>true</OptimizePNGs>
        <_XSAppIconAssets>Assets.xcassets/roslynpad.appiconset</_XSAppIconAssets>
      </PropertyGroup>
    ```
4. Add the following targets:
    ```xml
    <Target Name="Before_CoreCompileImageAssets" BeforeTargets="_CoreCompileImageAssets">
      <PropertyGroup>
        <Original_SdkPlatform>$(_SdkPlatform)</Original_SdkPlatform>
        <_SdkPlatform>MacCatalyst</_SdkPlatform>
        <Original_UIDeviceFamily>$(_UIDeviceFamily)</Original_UIDeviceFamily>
        <_UIDeviceFamily>IPad</_UIDeviceFamily>
        <Original_ComputedTargetFrameworkMoniker>$(_ComputedTargetFrameworkMoniker)</Original_ComputedTargetFrameworkMoniker>
        <_ComputedTargetFrameworkMoniker>$(_ComputedTargetFrameworkMoniker.Replace('=macos', '=maccatalyst'))</_ComputedTargetFrameworkMoniker>
      </PropertyGroup>
    </Target>
    <Target Name="After_CoreCompileImageAssets" AfterTargets="_CoreCompileImageAssets">
      <PropertyGroup>
        <_SdkPlatform>$(Original_SdkPlatform)</_SdkPlatform>
        <_UIDeviceFamily>$(Original_UIDeviceFamily)</_UIDeviceFamily>
        <_ComputedTargetFrameworkMoniker>$(Original_ComputedTargetFrameworkMoniker)</_ComputedTargetFrameworkMoniker>
      </PropertyGroup>
    </Target>
    ```

 A similar solution might work for icon bundling in other platforms.

**Comment by nixtar** [comment]
I've tried to reproduce this in one of my projects but I can't see any evidence that resizetizer is doing anything at all.

Should I see something in the build output/logs?

Do you also need to set the info.plist for the icon or does MauiIcon handle this?

**Reply by aelij** [reply]
Are you using the `net8.0-macos` target framework?
See a working example here - https://github.com/roslynpad/roslynpad/blob/main/src/RoslynPad.Avalonia/RoslynPad.Avalonia.csproj

**Reply by nixtar** [reply]
Sorry for late reply. Yes using `net8.0-macos` and I copied it from you RoslynPad repo exactly..

I ended up just doing it the hard way and manually creating the AppIcon.appiconset and referencing it in the Info.plist.
